feat(mobile): native_video_player

shenlong-tanwen opened the pull request:
#### Changes made in the PR:

- replaced video_player with native_video_player to fix HDR playback on mobile devices
- buffering is handled by using a timer to periodically check if the playback position has been updated or not

Fixes #2130
Fixes #3321
Fixes #5120
Fixes #5651
Fixes #7617
Fixes #11229
Fixes #11400
Fixes #11689

Maybe #5553? Need to test.

alextran1502 committed:
add native player library

alextran1502 committed:
splitup the player

alextran1502 committed:
stateful widget

shenlong-tanwen committed:
refactor: native_video_player

shenlong-tanwen committed:
fix: handle buffering

alextran1502 committed:
turn on volume when video plays

alextran1502 commented:
Good work as always!

A couple of issues I noticed

- After the video is uploaded, when viewing the video, it loads the remote video instead of the local one.
- The aspect ratio calculation for the remote video is not correct. It looks distorted
- When taping on the video from the timeline, you will often hear the sounds of the player getting initialized twice

hsu1943 commented:
Wow! Thanks a lot for your amazing work! Can this pull request fix issue #5120?

phajduk commented:
@shenlong-tanwen @alextran1502 any chance this PR will fix #11689  ?

alextran1502 commented:
@phajduk yes

phajduk commented:
@alextran1502 I will test this in a minute as I struggle with that bug hard and it's a bit of a showstopper to migrate to Immich.

alextran1502 commented:
@phajduk The pr is not finished yet, but yah, welcome to try out 😁

phajduk commented:
@alextran1502 @shenlong-tanwen photos has also wrong ratio (both local and remote). Attaching two images (one from Google Photos, another one from Immich).
![Screenshot_2024-08-30-21-45-26-17_965bbf4d18d205f782c6b8409c5773a4](https://github.com/user-attachments/assets/14ac9869-19a3-4f01-996a-266c989c784b)
![Screenshot_2024-08-30-21-44-58-57_d8aea95ef633eb6b7eb1cae13c34b40e](https://github.com/user-attachments/assets/b3cd4050-37d9-4051-ab31-712f1eb82374)

RaulGw4 commented:
Any chance this will also get fixed? #12006

shenlong-tanwen committed:
fix: aspect ratio

shenlong-tanwen commented:
@alextran1502 I cannot reproduce the issue where remote video is preferred instead of the local one and the audio issue. Found a possible root cause for the aspect ratio issue in case of local assets and handle it now. Can you please verify if you could reproduce the other issues still?

@phajduk Can you verify if you still face the aspect ratio issue with the recent change?

phajduk commented:
@shenlong-tanwen I've tested and 90% cases work. Thanks for doing this. I have some random private videos that have aspect ratio still broken but I can't upload it nor give specification because other, similar videos, work. I think we're good to merge for now.

dvbthien commented:
Hi @shenlong-tanwen 
When I tested the last commit, I encountered the same aspect ratio problem. I discovered that it occurs when I play remote videos that have HDR. Upon debugging, I noticed that asset.width and asset.height return incorrect values for HDR videos. And asset.width?.toDouble() doesn't return double value. 
<img width="418" alt="image" src="https://github.com/user-attachments/assets/5ab6983d-b763-4499-b7b9-331f6f71a246">
The dimensions of the video are 1080 pixels wide and pixels tall, but the resulting asset returns incorrect values.

alextran1502 commented:
I am also seeing errors when playing videos that were uploaded from iOS and play on Android (remote asset)

```
F/MPEG4Extractor(32196): frameworks/av/media/module/sec_extractors/mp4/MPEG4Extractor.cpp:7317 CHECK(AMediaFormat_getBuffer(format, AMEDIAFORMAT_KEY_CSD_HEVC, &data, &size)) failed.
V/MediaPlayerNative(30878): message received msg=100, ext1=100, ext2=1
E/MediaPlayerNative(30878): error (100, 1)
V/MediaPlayerNative(30878): message received msg=100, ext1=1, ext2=-2147483648
E/MediaPlayerNative(30878): error (1, -2147483648)
E/MediaPlayer(30878): Error (100,1)
D/VideoView(30878): Error: 100,1
E/MediaPlayer(30878): Error (1,-2147483648)
D/VideoView(30878): Error: 1,-2147483648
```

MagicLinus commented:
The playback issue with this HDR video significantly affects daily use. I'm eager for an upgrade to this player. Is there any progress on that? Thank you.

shenlong-tanwen committed:
fix: handle remote asset orientation

shenlong-tanwen commented:
The orientation issue should be fixed now.

poisonnuke commented:
As this PR is supposed to fix #5553 , please verify that all parts of the issue are fixed, that includes the BasicAuth issue on Android, ExoPlayer cannot play assets that use BasicAuth HTTP. 
See also #12985 for more informations. 

Im happy to provide an public accessible testing-URL on a private channel

alextran1502 commented:
@poisonnuke I don't think so

alextran1502 commented:
I've just finished another round of testing.

- On the emulator (iOS 17), the player can play the remote-only video. 
- On the physical device (iOS 18), the player cannot play the remote-only video. It can play the video that is on the device just fine. Here is a maybe relevant error message. 

```
#0      ChangeNotifier.debugAssertNotDisposed.<anonymous closure> (package:flutter/src/foundation/change_notifier.dart:183:9)
#1      ChangeNotifier.debugAssertNotDisposed (package:flutter/src/foundation/change_notifier.dart:190:6)
#2      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:416:27)
#3      ValueNotifier.value= (package:flutter/src/foundation/change_notifier.dart:559:5)
#4      NativeVideoViewerPage.build.initController (package:immich_mobile/pages/common/native_video_viewer.page.dart:214:18)
<asynchronous suspension>
```

Let me know if you need more information.

poisonnuke commented:
> @poisonnuke I don't think so

what does that mean? bo0tzz declared this as part of the bug and if this is the intended fix this PR must solve the BasicAuth issue. Therefore the BasicAuth test just became mandatory for this PR.

bo0tzz commented:
@poisonnuke 
1. You don't get to dictate how we should work
2. A PR doesn't have to fix every possible angle. If appropriate, the bug report or FR can stay open after the PR is merged.

rovo89 commented:
The ground work seems to be done though, so I assume there are chances to get Basic Authentication eventually: https://github.com/albemala/native_video_player/pull/25/files

Not sure there's also something for mTLS, which is what my family is eagerly waiting for.

alextran1502 commented:
@rovo89 neat, although it might be more applicable to be a subsequent pr since this pr is focusing on getting the player working first. There seems to be issue with playing remote videos on iOS 18 so we might have to split to use two players. So essentially the issue with the basic auth header won’t get resolved. But let’s not discussing that here to leave room for code related discussion ☺️

zackpollard commented:
> The ground work seems to be done though, so I assume there are chances to get Basic Authentication eventually: [albemala/native_video_player#25 (files)](https://github.com/albemala/native_video_player/pull/25/files)
> 
> Not sure there's also something for mTLS, which is what my family is eagerly waiting for.

Potentially, but this is not the aim of the PR in its current state, so lets keep discussion relevant to the PR for now and discussion around these improvements can stay in the existing github discussions :slightly_smiling_face:

rovo89 commented:
> it might be more applicable to be a subsequent pr

Sure - just wanted to give back some hope. 😉 And now I'll shut up and let you get this finalized.

zackpollard commented:
> > @poisonnuke
> > 
> > 1. You don't get to dictate how we should work
> 
> just to conclude this statement, this is wrong, you dictated it. Youve made a lot of mistakes that lead to this. The rest in private if youre really interessted in improving yourself. I was in your place 20 years ago, Ive made the same mistakes, too.

I'm not really sure what you're saying here. But this kind of input isn't appreciated and isn't valuable to the ongoing discussion. That statement is not wrong, you don't get to dictate how people work on their own open source projects, the level of entitlement here is astounding. Please enjoy continuing to use Immich, however we will no longer be accepting your input here. Have a nice day!

mertalev committed:
refinements and fixes

fix orientation for remote assets

wip separate widget

separate video loader widget

fixed memory leak

optimized seeking, cleanup

debug context pop

use global key

back to one widget

fixed rebuild

wait for swipe animation to finish

smooth hero animation for remote videos

faster scroll animation

mertalev committed:
clean up logging

mertalev committed:
refactor aspect ratio calculation

mertalev committed:
removed unnecessary import

mertalev commented:
I made a bunch of improvements to avoid animation stuttering / jank when opening or swiping a video, unnecessary async calls, repeated video initialization, etc. It works quite well in my testing at this point. However, there are quite a few changes I made to bring it to this point and I've only tested on iOS, so @alextran1502 please let me know if you notice any bugs. I also don't know how to fix the build failure for Android.

The only thing left on my mind is that seeking is not as responsive as I'd like, especially for remote videos. You have to let go or at least hold still for a moment for it to actually seek. This can probably be improved with better throttling / debouncing, but either way I don't think it's a blocker.

alextran1502 committed:
transitive dependencies

mertalev committed:
fixed referencing uninitialized orientation

mertalev commented:
I added throttling for the seek and it's a lot more responsive now. Not sure if 200ms delay is optimal, but it seems fine for now.

alextran1502 committed:
use correct ref to build android

mertalev committed:
higher res placeholder for local videos

mertalev committed:
slightly lower delay

mertalev committed:
await things

mertalev committed:
fix controls when swiping between image and video

mertalev committed:
linting

mertalev committed:
extra smooth seeking, add comments

mertalev commented:
I made a few more improvements, particularly with seeking. I'm pretty happy with where it's at for iOS. There's a small issue in that there's no audio in silent mode, which we may address in this PR or a later one. Android still needs to be tested.

Also, the PR will need to be rebased after #14086 is merged as there are a few changes that rely on it.

alextran1502 committed:
chore: generate router page

mertalev committed:
use current asset provider and loadAsset

mertalev committed:
fix stack handling

mertalev committed:
improved motion photo handling

mertalev committed:
use visibility for motion videos

mertalev committed:
error handling for async calls

mertalev committed:
fix duplicate key error

mertalev committed:
maybe fix duplicate key error

mertalev committed:
increase delay for hero animation

mertalev committed:
faster initialization for remote videos

mertalev committed:
ensure dimensions for memory cards

mertalev committed:
make aspect ratio logic reusable, optimizations

mertalev committed:
refactor: move exif search from aspect ratio to orientation

mertalev committed:
local orientation on ios is unreliable; prefer remote

mertalev committed:
fix no audio in silent mode on ios

mertalev committed:
increase bottom bar opacity to account for hdr

mertalev committed:
remove unused import

mertalev committed:
fix live photo play button not updating

mertalev committed:
fix map marker -> galleryviewer

mertalev committed:
remove video_player

mertalev committed:
fix hdr playback on android

mertalev committed:
fix looping

mertalev committed:
remove unused dependencies

mertalev committed:
update to latest player commit

mertalev committed:
fix player controls hiding when video is not playing

mertalev committed:
fix restart video

mertalev committed:
stop showing motion video after ending when looping is disabled

mertalev committed:
delay video initialization to avoid placeholder flicker

mertalev committed:
faster animation

mertalev committed:
shorter delay

mertalev committed:
small delay for image -> video on android

alextran1502 reviewed (approved):
Let's go 🚀

alextran1502 committed:
Merge branch 'main' into mobile/native-video-player

alextran1502 committed:
fix: lint

mertalev committed:
hide stacked children when controls are hidden, avoid bottom bar dropping

NicholasFlamy commented on mobile/lib/extensions/scroll_extensions.dart:
This fixed #5895! Thank you!
